File: cpp/codejam/y2020/QualificationRound/NestingDepth.cpp
/**
 * 2020.4.4
 *
 * 숫자로 이루어진 문자열이 있다.
 * 각 숫자는 깊이(depth)를 의미하고, 그 깊이에 맞게 괄호로 묶는 것이 문제.
 * 가장 최적으로(가장 짧게) 묶은 결과가 답이다.
 *
 * (input)
 * T: 테스트케이스의 수
 * S: 숫자로 구성된 문자열. Test set 1은 0과 1, Test set 2는 0부터 9.
 *
 * (output)
 * 각 숫자를 깊이에 맞게 괄호로 묶어준 결과
 * 0 -> 0, 1 -> (1), 2 -> ((2)), 012 -> 0(1(2)), ...
 *
 * (solution 1)
 * S를 앞에서부터 탐색
 * "현재깊이"와 비교하면서 같으면 넘어가고,
 * 현재깊이보다 크면 괄호 열고, 작으면 괄호 닫고.
 *
 * (solution 1 result)
 * 성공!!
 */

#include <stdio.h>
#include <iostream>
#include <sstream>
#include <algorithm>

using namespace std;

string S;

string solve() {
	stringstream ss;
	int curr_depth = 0;
	for (int i=0; i<S.size(); i++) {
		int n = S[i] - '0';
		if (n > curr_depth) {
			int size = n - curr_depth;
			for (int j=0; j<size; j++) {
				ss << "(";
				curr_depth++;
			}
		} else if (n < curr_depth) {
			int size = curr_depth - n;
			for (int j=0; j<size; j++) {
				ss << ")";
				curr_depth--;
			}
		}
		ss << S[i];
	}
	for (int i=0; i<curr_depth; i++) {
		ss << ")";
	}
	return ss.str();
}

int main() {
	int T;
	cin >> T;
	for (int i=0; i<T; i++) {
		cin >> S;
		string result = solve();
		cout << "Case #" << i+1 << ": " << result << endl;
	}
}

File: python/codejam/y2020/qualification_round/nesting_depth.py
'''
2021.3.13

0~9 숫자로 구성된 문자열 S에서, 
각 숫자 digit들이 그 수만큼의 깊이로 괄호안에 들어가도록 괄호가 추가된 문자열 만들어서 리턴.
01221 => 0(1(22)1)

(input)
T: 테스트케이스의 수
S: 0~9로 구성된 문자열

(output)
Case #x: ss
ss: S에 규칙에 맞도록 괄호 추가된 문자열

(solution 1)
S의 각 digit을 탐색하면서, ss를 만드는데, 
괄호 열린 숫자 p_open과 digit을 비교하여 여는 괄호 및 닫는 괄호를 추가한다.
'''

def solve(S):
    ss = ''
    p_open = 0
    for dd in S:
        d = int(dd)
        if d > p_open:
            ss += '(' * (d - p_open)
        elif d < p_open:
            ss += ')' * (p_open - d)
        p_open = d
        ss += dd
    ss += ')' * p_open
    return ss

T = int(input())
for tt in range(1, T+1):
    S = input()
    result = solve(S)
    print('Case #{}: {}'.format(tt, result))